File: plantbreed/pkg/R/diallele1.r
diallele1 <- function (dataframe,yvar = "yvar", progeny = "progeny",  male = "male", female = "female",  replication = "replication")
{
dataframe <- data.frame (yvar = dataframe[,yvar],progeny = dataframe[,progeny],male = dataframe[, male], female = dataframe[,female], replication = dataframe[, replication])
 dataframe$male = as.factor(dataframe$male)
dataframe$female = as.factor(dataframe$female)
dataframe$progeny = as.factor(dataframe$progeny)
dataframe$replication =  as.factor(dataframe$replication)
mean.y =  mean(as.numeric(dataframe$yvar))

#Analysis of variance
formula = paste ("yvar"," ~ ", "progeny", " + ", "replication")
md1 <- lm(formula, data= dataframe) # anova treatment
yvar = yvar
 cat("Diallel analysis for trait: ", yvar, "\n\n")
 cat("...........................\n\n")
anvout <- anova(md1)
print(anvout)
if ( anvout[1,5] > 0.05){
           cat ("Note: Progeny effect is not significant at 0.05 p threshold")
           }
if ( anvout[1,5] > 0.01){
           cat ("Note: Progeny effect is not significant at 0.01 p threshold")
           }

data1 <- data.frame(aggregate(yvar ~ male: female, data = dataframe, mean))

mydf <- aggregate(yvar ~ female, data1, 'c')

myMatrix <- as.matrix (mydf[,-1])

n = (length(myMatrix))^0.5

acon <- sum((1/(2 * n))*((rowSums(myMatrix) + colSums(myMatrix))^2))
ssgca = acon - (2/ (n ^2))* (sum(myMatrix)^2)

sssca  = sum((1/ 2)*(myMatrix * (myMatrix + t(myMatrix)))) - acon + (1/ (n ^2))* (sum(myMatrix)^2)

ssrecp =   ((1/ 4)* sum((myMatrix - t(myMatrix))^2))
 r = nlevels(as.factor(dataframe$replication))
MSEAD = anvout$`Mean Sq`[3] / r

p = n # number of parents
Df <- c((p-1), (p *(p-1)/2), (p*(p-1)/2), anvout$Df[3])
SSS <- c(ssgca, sssca,ssrecp)
ssq <- c( SSS, anvout$`Sum Sq`[3]/r)
MSSS = SSS/Df[1:3]
MSSS1 = c(MSSS,MSEAD)
FVAL = c(MSSS1[1:3]/MSEAD, NA)
pval = c(1- pf(FVAL[1:3], Df[1:3], Df[3]), NA)
anovadf.mod1 <- data.frame(Df, `Sum Sq`=ssq, `Mean Sq`=MSSS1, `F value`= FVAL, `Pr(>F)`= pval, check.names=FALSE)
rownames(anovadf.mod1) <- c("GCA","SCA","Reciprocal", "error")
class(anovadf.mod1) <- c("anova","data.frame")
yvar <- names(dataframe$yvar)
yvar <- yvar
cat("Anova for combining ability - ModelI", yvar, "\n\n")
print(anovadf.mod1)

# genetic components
#GCA component
 GCAcomp = (MSSS[1] - MSEAD) / (2 * n)
 # SCA component

  SCAcomp = (MSSS[2] - MSEAD)
  RCAcomp  = (MSSS[3] - MSEAD) / 2
  GCARCAratio = GCAcomp /  SCAcomp
 components.model1 <- list (GCAcomp = GCAcomp, SCAcomp = SCAcomp, RCAcomp = RCAcomp, GCRCAratio = GCARCAratio)

  cat("Componets: Model 1",  "\n\n")
  cat("GCA :", GCAcomp, "\n\n")
  cat("SCA :", SCAcomp, "\n\n")
  cat("Reciprocal:", RCAcomp, "\n\n")
  cat("GCA to RCA ratio: ", GCARCAratio, "\n\n")

# GCA effects
gcaeff <- ((1/(2 * n))*(rowSums(myMatrix) + colSums(myMatrix))) - ((1/ (n ^2))* (sum(myMatrix)))
scaeff <- ((1/ 2) * (myMatrix + t(myMatrix))) -  ((1/(2 *n ))* (rowSums(myMatrix) + colSums(myMatrix) + colSums(t(myMatrix)) + rowSums(t(myMatrix)))) + ((1/ (n ^2))* (sum(myMatrix)))
recieff <-  0.5 * (myMatrix - t(myMatrix))
effmat <- gcaeff
#effmat <- diag(gcaeff)
#effmat <- lower.tri(recieff, diag = FALSE)

# Variances: standard error and critical differences
var.gi = ((n-1) / (2 * n ^2)) * MSEAD
var.sii = (((n-1)^2) / n ^2 ) *  MSEAD
var.sij = (1/(2*n^2))*((n^2)-2*n + 2) * MSEAD
var.rij =  0.5 * MSEAD
var.gi_gj = (1/n)*MSEAD
var.sij_sji = ((2*(n-2)) / n)* MSEAD
var.sii_sij = ((3*n -2)/ (2*n)) * MSEAD
var.sii_sjk =  ((3*(n-2)) / (2*n))* MSEAD
var.sij_sik = ((n-1)/n )* MSEAD
var.sij_skl = ((n-2)/n) * MSEAD
var.rij_rkl = MSEAD

varcompare <- list ( var.gi = var.gi,var.sii = var.sii, var.sij = var.sij, var.rij = var.rij, var.gi_gj = var.gi_gj, var.sij_sji = var.sij_sji,
var.sii_sij = var.sii_sij, var.sii_sjk = var.sii_sjk, var.sij_sik = var.sij_sik, var.sij_skl = var.sij_skl, var.rij_rkl = var.rij_rkl)

# model II
# Test of significance
cat("Anova for combining ability, model II", yvar, "\n\n")
FVAL1 = c(MSSS1[1]/MSSS1[2],MSSS1[2:3]/MSEAD, NA)
pval1 = c(1- pf(FVAL1[1], Df[1], Df[2]),1- pf(FVAL1[2:3], Df[2:3], Df[4]), NA)
anovadf.mod2 <- data.frame(Df, `Sum Sq`=ssq, `Mean Sq`=MSSS1, `F value`= FVAL1, `Pr(>F)`= pval1, check.names=FALSE)
rownames(anovadf.mod2) <- c("GCA","SCA","Reciprocal", "error")
class(anovadf.mod2) <- c("anova","data.frame")
print(anovadf.mod2)
# genetic components estimates
sigmasq.g = (1/(2* n))*(MSSS1[1] - (((MSEAD + n * (n-1) * MSSS1[2])) / (n^2 - n + 1)))
sigmasq.s = ((n ^ 2) / (2*(n^2 - n + 1))) * (MSSS1[2] - MSEAD )
sigmasq.r = 0.5 * (MSSS1[3] - MSEAD )
sigmasq.error = MSEAD
sgimasq.A = 2 * sigmasq.g
sigmasq.D = sigmasq.s
gca.scaratio = sigmasq.g /sigmasq.s

  cat("Componets: Model 2",  "\n\n")
  cat("GCA :", sigmasq.g , "\n\n")
  cat("SCA :", sigmasq.s, "\n\n")
  cat("Reciprocal:", sigmasq.r, "\n\n")
  cat("GCA to RCA ratio: ", gca.scaratio, "\n\n")

varcomp.model2 = list (sigmasq.g = sigmasq.g, sigmasq.s = sigmasq.s, sigmasq.r = sigmasq.r ,
sigmasq.error = sigmasq.error,sgimasq.A = sgimasq.A, sigmasq.D = sigmasq.D, gca.scaratio = gca.scaratio)

 # need to be expanded
return (list (anvout = anvout, anova.mod1 = anovadf.mod1,components.model1 = components.model1, gca.effmat = gcaeff, sca.effmat = scaeff,
reciprocal.effmat = recieff, varcompare = varcompare,anovadf.mod2 = anovadf.mod2,
varcomp.model2 = varcomp.model2))
}

File: core_logic/python_package/plantbreeding_py/diallele1.py
import pandas as pd
import numpy as np
import statsmodels.api as sm
import statsmodels.formula.api as smf
from scipy.stats import f

def diallele1(dataframe, yvar="yvar", progeny="progeny", 
              male="male", female="female", 
              replication="replication", verbose=True):
    """
    Analysis of Diallel data

    Calculates general and specific combining ability and other estimates for 
    Diallel mating design using Griffing's Method I (parents, F1s, and reciprocals) 
    (Griffing, 1956). 
    """
    
    # 1. Input Validation and Data Preparation
    req_cols = [yvar, progeny, male, female, replication]
    if not all(col in dataframe.columns for col in req_cols):
        raise ValueError("One or more specified columns are not found in the dataframe.")
    
    # Work on a clean subset
    df = dataframe[req_cols].copy()
    
    # Ensure categorical variables are treated as factors (strings/objects)
    for col in [progeny, male, female, replication]:
        df[col] = df[col].astype(str)
        
    df[yvar] = pd.to_numeric(df[yvar])
    
    mean_y = df[yvar].mean()
    
    # 2. Initial Analysis of Variance (Treatments)
    # Using statsmodels to fit the linear model
    formula_str = f"{yvar} ~ C({progeny}) + C({replication})"
    md1 = smf.ols(formula_str, data=df).fit()
    anvout = sm.stats.anova_lm(md1, typ=1)
    
    # Identify the residual index (statsmodels usually calls it 'Residual')
    resid_idx = "Residual" if "Residual" in anvout.index else anvout.index[-1]
    
    if verbose:
        print(f"Diallel analysis for trait: {yvar}")
        print("...........................\n")
        print(anvout)
        
        progeny_p = anvout.loc[f"C({progeny})", "PR(>F)"]
        if progeny_p > 0.05:
            print("\nNote: Progeny effect is not significant at 0.05 p-threshold\n")
        elif progeny_p > 0.01:
            print("\nNote: Progeny effect is not significant at 0.01 p-threshold\n")
            
    # 3. Create the Male x Female Matrix using a pivot table
    pivot_df = df.pivot_table(values=yvar, index=male, columns=female, aggfunc='mean')
    myMatrix = pivot_df.values
    parents = pivot_df.index.tolist()
    
    if myMatrix.shape[0] != myMatrix.shape[1]:
        raise ValueError("The number of male and female parents do not match. Matrix is not square.")
        
    n = myMatrix.shape[0]
    
    # 4. Sums of Squares Calculations
    row_sums = myMatrix.sum(axis=1)
    col_sums = myMatrix.sum(axis=0)
    matrix_sum = myMatrix.sum()
    
    acon = np.sum((1 / (2 * n)) * ((row_sums + col_sums)**2))
    ssgca = acon - (2 / (n**2)) * (matrix_sum**2)
    sssca = np.sum((1 / 2) * (myMatrix * (myMatrix + myMatrix.T))) - acon + (1 / (n**2)) * (matrix_sum**2)
    ssrecp = (1 / 4) * np.sum((myMatrix - myMatrix.T)**2)
    
    r = df[replication].nunique()
    MSEAD = anvout.loc[resid_idx, "mean_sq"] / r
    
    # 5. Combining Ability ANOVA - Model I (Fixed)
    Df = [n - 1, n * (n - 1) / 2, n * (n - 1) / 2, anvout.loc[resid_idx, "df"]]
    SSS = [ssgca, sssca, ssrecp]
    ssq = SSS + [anvout.loc[resid_idx, "sum_sq"] / r]
    
    MSSS = [SSS[i] / Df[i] for i in range(3)]
    MSSS1 = MSSS + [MSEAD]
    
    FVAL = [MSSS1[0] / MSEAD, MSSS1[1] / MSEAD, MSSS1[2] / MSEAD, np.nan]
    pval = [f.sf(FVAL[i], Df[i], Df[3]) if not np.isnan(FVAL[i]) else np.nan for i in range(4)]
    
    anovadf_mod1 = pd.DataFrame({
        "Df": Df,
        "Sum Sq": ssq,
        "Mean Sq": MSSS1,
        "F value": FVAL,
        "Pr(>F)": pval
    }, index=["GCA", "SCA", "Reciprocal", "Error"])
    
    if verbose:
        print("\nAnova for combining ability - Model I (Fixed)")
        print(anovadf_mod1)
        
    # Genetic Components - Model I
    GCAcomp = (MSSS[0] - MSEAD) / (2 * n)
    SCAcomp = (MSSS[1] - MSEAD)
    RCAcomp = (MSSS[2] - MSEAD) / 2
    GCARCAratio = GCAcomp / SCAcomp if SCAcomp != 0 else np.nan
    
    components_model1 = {
        "GCAcomp": GCAcomp, 
        "SCAcomp": SCAcomp, 
        "RCAcomp": RCAcomp, 
        "GCRCAratio": GCARCAratio
    }
    
    if verbose:
        print("\nComponents: Model 1")
        print(f"GCA : {GCAcomp}")
        print(f"SCA : {SCAcomp}")
        print(f"Reciprocal: {RCAcomp}")
        print(f"GCA to SCA ratio: {GCARCAratio}")
        
    # 6. GCA, SCA, and Reciprocal Effects Matrices
    P = row_sums + col_sums
    
    gcaeff = ((1 / (2 * n)) * P) - ((1 / (n**2)) * matrix_sum)
    gcaeff_series = pd.Series(gcaeff, index=parents)
    
    # Utilizing np.newaxis to correctly broadcast the parent additions (Xi. + X.i + Xj. + X.j)
    scaeff = (0.5 * (myMatrix + myMatrix.T)) - \
             ((1 / (2 * n)) * (P[:, np.newaxis] + P[np.newaxis, :])) + \
             ((1 / (n**2)) * matrix_sum)
    scaeff_df = pd.DataFrame(scaeff, index=parents, columns=parents)
             
    recieff = 0.5 * (myMatrix - myMatrix.T)
    recieff_df = pd.DataFrame(recieff, index=parents, columns=parents)
    
    # Variances: standard error and critical differences
    varcompare = {
        "var_gi": ((n - 1) / (2 * n**2)) * MSEAD,
        "var_sii": (((n - 1)**2) / n**2) * MSEAD,
        "var_sij": (1 / (2 * n**2)) * ((n**2) - 2 * n + 2) * MSEAD,
        "var_rij": 0.5 * MSEAD,
        "var_gi_gj": (1 / n) * MSEAD,
        "var_sij_sji": ((2 * (n - 2)) / n) * MSEAD,
        "var_sii_sij": ((3 * n - 2) / (2 * n)) * MSEAD,
        "var_sii_sjk": ((3 * (n - 2)) / (2 * n)) * MSEAD,
        "var_sij_sik": ((n - 1) / n) * MSEAD,
        "var_sij_skl": ((n - 2) / n) * MSEAD,
        "var_rij_rkl": MSEAD
    }
    
    # 7. Combining Ability ANOVA - Model II (Random)
    FVAL1 = [MSSS1[0] / MSSS1[1], MSSS1[1] / MSEAD, MSSS1[2] / MSEAD, np.nan]
    pval1 = [
        f.sf(FVAL1[0], Df[0], Df[1]),
        f.sf(FVAL1[1], Df[1], Df[3]),
        f.sf(FVAL1[2], Df[2], Df[3]),
        np.nan
    ]
    
    anovadf_mod2 = pd.DataFrame({
        "Df": Df,
        "Sum Sq": ssq,
        "Mean Sq": MSSS1,
        "F value": FVAL1,
        "Pr(>F)": pval1
    }, index=["GCA", "SCA", "Reciprocal", "Error"])
    
    if verbose:
        print("\nAnova for combining ability - Model II (Random)")
        print(anovadf_mod2)
        
    # Genetic components estimates - Model II
    sigmasq_g = (1 / (2 * n)) * (MSSS1[0] - (((MSEAD + n * (n - 1) * MSSS1[1])) / (n**2 - n + 1)))
    sigmasq_s = ((n**2) / (2 * (n**2 - n + 1))) * (MSSS1[1] - MSEAD)
    sigmasq_r = 0.5 * (MSSS1[2] - MSEAD)
    sigmasq_error = MSEAD
    sigmasq_A = 2 * sigmasq_g
    sigmasq_D = sigmasq_s
    gca_scaratio = sigmasq_g / sigmasq_s if sigmasq_s != 0 else np.nan
    
    if verbose:
        print("\nComponents: Model 2")
        print(f"GCA (sigma^2 g) : {sigmasq_g}")
        print(f"SCA (sigma^2 s) : {sigmasq_s}")
        print(f"Reciprocal (sigma^2 r): {sigmasq_r}")
        print(f"GCA to SCA ratio: {gca_scaratio}\n")
        
    varcomp_model2 = {
        "sigmasq_g": sigmasq_g, 
        "sigmasq_s": sigmasq_s, 
        "sigmasq_r": sigmasq_r, 
        "sigmasq_error": sigmasq_error,
        "sigmasq_A": sigmasq_A, 
        "sigmasq_D": sigmasq_D, 
        "gca_scaratio": gca_scaratio
    }
    
    # 8. Return comprehensive dictionary
    results = {
        "anvout": anvout, 
        "anova_mod1": anovadf_mod1,
        "components_model1": components_model1, 
        "gca_effmat": gcaeff_series, 
        "sca_effmat": scaeff_df,
        "reciprocal_effmat": recieff_df, 
        "varcompare": varcompare,
        "anova_mod2": anovadf_mod2,
        "varcomp_model2": varcomp_model2
    }
    
    return results
